solr/core/src/java/org/apache/solr/core/SolrCores.java

```diff
@@ -51,7 +51,7 @@
   // to essentially queue them up to be handled via pendingCoreOps.
   private static final List<SolrCore> pendingCloses = new ArrayList<>();
 
-  private TransientSolrCoreCacheFactory transientSolrCoreCacheFactory;
+  private TransientSolrCoreCacheFactory transientSolrCoreCacheFactory = TransientSolrCoreCacheFactory.NO_OP;
```

Comment:
Under what circumstance do we need this no-op impl to prevent an NPE?

Reply:
If a SolrCores method accessing getTransientCacheHandler() is called before calling load().
For example an asynchronous periodic thread that would be started with CoreContainer and that would call getLoadedCoreNames(). This worked before even if SolrCores.load() was called after. (first calls would return without counting transient cores, and subsequent calls after load() would count transient cores)

Reply:
then SolrCores.load could be called much sooner, basically right after the resourceLoader is ready.  

Also, maybe this other thread ought to wait to start till some later time.  Perhaps ideally there would be an event publishing mechanism, which doesn't exist currently, I know.  Or alternatively just have some CountDownLatch signal, like signaling when Solr will begin loading cores.